File: typescript/examples/mainnet.ts
/**
 * Example: Query NEAR mainnet for various information
 * 
 * Run with: bun run examples/mainnet.ts
 */

import { NearRpcClient } from "../src/client.js";

async function main() {
  const client = NearRpcClient.mainnet();

  console.log("=== NEAR Mainnet RPC Client Test (TypeScript) ===\n");

  // 1. Get node status
  console.log("1. Fetching node status...");
  const status = await client.status();
  console.log(`   Chain ID: ${status.chain_id}`);
  console.log(`   Protocol version: ${status.protocol_version}`);
  console.log(`   Latest block height: ${status.sync_info.latest_block_height}`);
  console.log(`   Latest block hash: ${status.sync_info.latest_block_hash}`);
  console.log(`   Syncing: ${status.sync_info.syncing}`);
  console.log();

  // 2. Get the latest block
  console.log("2. Fetching latest block...");
  const block = await client.block({ finality: "final" });
  console.log(`   Block height: ${block.header.height}`);
  console.log(`   Block hash: ${block.header.hash}`);
  console.log(`   Timestamp: ${block.header.timestamp}`);
  console.log(`   Author: ${block.author}`);
  console.log(`   Chunks: ${block.chunks.length}`);
  console.log();

  // 3. Get gas price
  console.log("3. Fetching gas price...");
  const gas = await client.gasPrice({ block_id: null });
  console.log(`   Gas price: ${gas.gas_price} yoctoNEAR`);
  console.log();

  // 4. Query account
  console.log("4. Querying 'near' foundation account...");
  const accountQuery = await client.query({
    request_type: "view_account",
    finality: "final",
    account_id: "near",
  });
  
  // The response type depends on request_type
  if ("amount" in accountQuery) {
    const balanceYocto = BigInt(accountQuery.amount);
    const balanceNear = balanceYocto / BigInt(10 ** 24);
    console.log(`   Account: near`);
    console.log(`   Balance: ${balanceNear} NEAR`);
    console.log(`   Storage used: ${accountQuery.storage_usage} bytes`);
  }
  console.log();

  // 5. Call a view function on a contract
  console.log("5. Calling view function on wrap.near...");
  const callResult = await client.query({
    request_type: "call_function",
    finality: "final",
    account_id: "wrap.near",
    method_name: "ft_metadata",
    args_base64: btoa("{}"), // base64 encode "{}"
  });

  if ("result" in callResult && Array.isArray(callResult.result)) {
    const decoded = new TextDecoder().decode(new Uint8Array(callResult.result));
    console.log(`   wNEAR metadata: ${decoded}`);
  }
  console.log();

  // 6. Get validators
  console.log("6. Fetching current validators...");
  const validators = await client.validators({ latest: null });
  console.log(`   Current validators: ${validators.current_validators.length}`);
  console.log(`   Next validators: ${validators.next_validators.length}`);
  if (validators.current_validators.length > 0) {
    const first = validators.current_validators[0];
    console.log(`   Top validator: ${first.account_id} (stake: ${first.stake})`);
  }
  console.log();

  // 7. Network info
  console.log("7. Fetching network info...");
  const network = await client.networkInfo();
  console.log(`   Active peers: ${network.active_peers.length}`);
  console.log(`   Known producers: ${network.known_producers.length}`);
  console.log();

  console.log("=== All tests passed! ===");
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: rust/examples/mainnet.rs
//! Example: Query NEAR mainnet for various information.
//!
//! Run with: cargo run --example mainnet

use near_rpc_client::{client::Result, types::*, NearRpcClient};

#[tokio::main]
async fn main() -> Result<()> {
    let client = NearRpcClient::mainnet();

    println!("=== NEAR Mainnet RPC Client Test ===\n");

    // 1. Get node status
    println!("1. Fetching node status...");
    let status = client.status().await?;
    println!("   Chain ID: {}", status.chain_id);
    println!("   Protocol version: {}", status.protocol_version);
    println!(
        "   Latest block height: {}",
        status.sync_info.latest_block_height
    );
    println!(
        "   Latest block hash: {}",
        status.sync_info.latest_block_hash
    );
    println!("   Syncing: {}", status.sync_info.syncing);
    println!();

    // 2. Get the latest block (using finality)
    println!("2. Fetching latest block...");
    let block = client
        .block(RpcBlockRequest::Finality(Finality::Final))
        .await?;
    println!("   Block height: {}", block.header.height);
    println!("   Block hash: {}", block.header.hash);
    println!("   Timestamp: {}", block.header.timestamp);
    println!("   Author: {}", block.author);
    println!("   Chunks: {}", block.chunks.len());
    println!();

    // 3. Get gas price (null block_id returns latest)
    println!("3. Fetching gas price...");
    let gas = client
        .gas_price(RpcGasPriceRequest { block_id: None })
        .await?;
    println!("   Gas price: {} yoctoNEAR", gas.gas_price);
    println!();

    // 4. Query account - using cartesian product variant
    println!("4. Querying 'near' foundation account...");
    let account_query = client
        .query(RpcQueryRequest::ViewAccountFinality {
            finality: Finality::Final,
            account_id: "near".parse().expect("valid account id"),
            request_type: "view_account".to_string(),
        })
        .await?;

    if let RpcQueryResponse::AccountView {
        amount,
        storage_usage,
        ..
    } = account_query
    {
        // amount is NearToken (u128 as string)
        let balance_str: String = amount.into();
        let balance_yocto: u128 = balance_str.parse().unwrap_or(0);
        let balance_near = balance_yocto / 10u128.pow(24);
        println!("   Account: near");
        println!("   Balance: {} NEAR", balance_near);
        println!("   Storage used: {} bytes", storage_usage);
    }
    println!();

    // 5. Call a view function on a contract
    println!("5. Calling view function on wrap.near...");
    let call_result = client
        .query(RpcQueryRequest::CallFunctionFinality {
            finality: Finality::Final,
            account_id: "wrap.near".parse().expect("valid account id"),
            method_name: "ft_metadata".to_string(),
            args_base64: "e30=".parse().expect("valid base64"), // base64 encoded "{}"
            request_type: "call_function".to_string(),
        })
        .await?;

    if let RpcQueryResponse::CallResult { result, .. } = call_result {
        let decoded = String::from_utf8(result).unwrap_or_default();
        println!("   wNEAR metadata: {}", decoded);
    }
    println!();

    // 6. Get validators (using Latest variant)
    println!("6. Fetching current validators...");
    let validators = client.validators(RpcValidatorRequest::Latest).await?;
    println!(
        "   Current validators: {}",
        validators.current_validators.len()
    );
    println!("   Next validators: {}", validators.next_validators.len());
    if let Some(first) = validators.current_validators.first() {
        println!(
            "   Top validator: {} (stake: {})",
            first.account_id, first.stake
        );
    }
    println!();

    // 7. Network info
    println!("7. Fetching network info...");
    let network = client.network_info().await?;
    println!("   Active peers: {}", network.active_peers.len());
    println!("   Known producers: {}", network.known_producers.len());
    println!();

    println!("=== All tests passed! ===");
    Ok(())
}
